File: genetic_algorithm_ruby/lib/bio_genome.rb
class Genome
    def initialize
        
    end
    
    def generate_ssDNA(length) # single strand DNA
        nucleotide = ["A", "T", "G", "C"]
        chromosome = []
        until chromosome.length >= length
        chromosome << nucleotide.sample
        end
        chromosome.join
    end

    def generate_rna(length)
        nucleotide = ["A", "U", "G", "C"]
        chromosome = []
        until chromosome.length >= length
        chromosome << nucleotide.sample
        end
        chromosome.join
    end

    def complementary_DNA_strand(dna)
        dna = dna.upcase
        for i in 0..(dna.size - 1)
            case dna[i]
            when "A"
                dna[i] = "T"
            when "T"
                dna[i] = "A"
            when "G"
                dna[i] = "C"
            when "C" 
                dna[i] = "G"
            end
        end
        dna
    end

    def check_DNA(seq1, seq2)
        if seq1.length >= seq2.length
    
            strand2 = seq2.reverse
            strand2_check = seq1.tr("ACTG", "TGAC")
            strand2_check.include? strand2
        else
            strand1 = seq1.reverse
            strand1_check = seq2.tr("ACTG", "TGAC")
            strand1_check.include? strand1
        end
    end

    def six_reading_frames(seq1, rseq2)
        output = []
        
        frame1 = seq1.scan(/.../).join" "
        f2 = seq1[1..-3].scan(/.../).join" "
        frame2 = seq1[0] + " " + f2 + " " + seq1[-2..-1]
        f3 = seq1[2..-2].scan(/.../).join" "
        frame3 = seq1[0..1] + " " + f3 + " " + seq1[-1]
        output << [frame1, f2, f3].join("\n")
        # output << "\n"
    
        
        rframe1 = rseq2.scan(/.../).join" "
        f2 = rseq2[1..-3].scan(/.../).join" "
        rframe2 = rseq2[0] + " " + f2 + " " + rseq2[-2..-1]
        f3 = rseq2[2..-2].scan(/.../).join" "
        rframe3 = rseq2[0..1] + " " + f3 + " " + rseq2[-1]
        output << [rframe1, f2, f3].join("\n")
      
        return output
    end
    
    def translate_with_frame(dna="TGCGATGAATGGGCTCGCTCC", frames=[1,2,3,-1,-2,-3])
        aminoacid_dictionary = {'UUC'=>'F', 'UUU'=>'F','UUA'=>'L', 'UUG'=>'L', 'CUU'=>'L', 'CUC'=>'L','CUA'=>'L','CUG'=>'L', 'AUU'=>'I', 'AUC'=>'I', 'AUA'=>'I','AUG'=>'M','GUU'=>'V', 'GUC'=>'V', 'GUA'=>'V', 'GUG'=>'V', 'UCU'=>'S', 'UCC'=>'S', 'UCA'=>'S', 'UCG'=>'S', 'AGU'=>'S', 'AGC'=>'S', 'CCU'=>'P', 'CCC'=>'P', 'CCA'=>'P', 'CCG'=>'P', 'ACU'=>'T', 'ACC'=>'T', 'ACA'=>'T', 'ACG'=>'T', 'GCU'=>'A', 'GCC'=>'A', 'GCA'=>'A', 'GCG'=>'A', 'UAU'=>'Y', 'UAC'=>'Y', 'CAU'=>'H', 'CAC'=>'H', 'CAA'=>'Q', 'CAG'=>'Q', 'AAU'=>'N', 'AAC'=>'N', 'AAA'=>'K', 'AAG'=>'K', 'GAU'=>'D', 'GAC'=>'D', 'GAA'=>'E', 'GAG'=>'E', 'UGU'=>'C', 'UGC'=>'C', 'UGG'=>'W', 'CGU'=>'R', 'CGC'=>'R', 'CGA'=>'R', 'CGG'=>'R', 'AGA'=>'R', 'AGG'=>'R', 'GGU'=>'G',  'GGC'=>'G', 'GGA'=>'G', 'GGG'=>'G', 'UAA'=>'*', 'UGA'=>'*', 'UAG'=>'*'}
        
        output = []
    
        if dna == ""
            for i in 0...frames.length
            output << ""
            end
        return output
        end
    
       
        frame1 = dna.gsub('T','U').scan(/.../).join""
        frame2 = dna.gsub('T','U')[1..-3].scan(/.../).join""
        frame3 = dna.gsub('T','U')[2..-2].scan(/.../).join""
      
        rframe1 = dna.tr("ACGT", "TGCA").reverse.gsub('T','U').scan(/.../).join""
        rframe2 = dna.tr("ACGT", "TGCA").reverse.gsub('T','U')[1..-3].scan(/.../).join""
        rframe3 = dna.tr("ACGT", "TGCA").reverse.gsub('T','U')[2..-2].scan(/.../).join""
      
        output << [frame1, frame2, frame3, rframe1,rframe2, rframe3] 
    
        
        translations = output.flatten.map do |frames| 
            frames.scan(/.../).map{|match|aminoacid_dictionary[match]}.join 
        end
    
        frames.map do |frame| 
            case
                when frame == -1 then frame = 4
                when frame == -2 then frame = 5
                when frame == -3 then frame = 6
                else 
                    ""
            end
        translations[frame-1]
        end
        
    end

    def dna_to_rna(dna = "GATCTCCGTC")
        #your code here
        for i in 0..(dna.size - 1)
          # if i == "T"
          dna[i] = dna[i] == "T" ? "U" : dna[i]
        end
        dna
    end

    $aminoacid_dictionary = {'UUC'=>'F', 'UUU'=>'F','UUA'=>'L', 'UUG'=>'L', 'CUU'=>'L', 'CUC'=>'L','CUA'=>'L','CUG'=>'L', 'AUU'=>'I', 'AUC'=>'I', 'AUA'=>'I','AUG'=>'M','GUU'=>'V', 'GUC'=>'V', 'GUA'=>'V', 'GUG'=>'V', 'UCU'=>'S', 'UCC'=>'S', 'UCA'=>'S', 'UCG'=>'S', 'AGU'=>'S', 'AGC'=>'S', 'CCU'=>'P', 'CCC'=>'P', 'CCA'=>'P', 'CCG'=>'P', 'ACU'=>'T', 'ACC'=>'T', 'ACA'=>'T', 'ACG'=>'T', 'GCU'=>'A', 'GCC'=>'A', 'GCA'=>'A', 'GCG'=>'A', 'UAU'=>'Y', 'UAC'=>'Y', 'CAU'=>'H', 'CAC'=>'H', 'CAA'=>'Q', 'CAG'=>'Q', 'AAU'=>'N', 'AAC'=>'N', 'AAA'=>'K', 'AAG'=>'K', 'GAU'=>'D', 'GAC'=>'D', 'GAA'=>'E', 'GAG'=>'E', 'UGU'=>'C', 'UGC'=>'C', 'UGG'=>'W', 'CGU'=>'R', 'CGC'=>'R', 'CGA'=>'R', 'CGG'=>'R', 'AGA'=>'R', 'AGG'=>'R', 'GGU'=>'G',  'GGC'=>'G', 'GGA'=>'G', 'GGG'=>'G', 'UAA'=>'Stop', 'UGA'=>'Stop', 'UAG'=>'Stop'}
    def protein(rna)
        rna.upcase.scan(/.../).take_while{|s| s }.map{|match|$aminoacid_dictionary[match]}.join
        # != 'UAA' && s != 'UGA' && s != 'UAG'
    end

    def run()
    end

end


genome = Genome.new

dna = genome.generate_ssDNA(10)
# p rna
dna = "TGCGATGAATGGGCTCGCTCC"
puts "DNA is #{dna}"
cdna = genome.complementary_DNA_strand(dna)
puts "the complementary strand is #{cdna}" 
rcdna = cdna.reverse
checked = genome.check_DNA(dna,rcdna)
puts "The complementary strand, when arranged 5 prime to 3 prime, is '#{rcdna}' , confirmed to be #{checked}"

mrna = genome.dna_to_rna(dna)
puts "when the dna undergoes transcription, the mrna reads #{mrna}"
# proteins = genome.protein(mrna) 
# puts "When translated it is #{proteins}"
  
# # frames = genome.six_reading_frames(dna, rcdna)
# # puts "the dna read as 6 frames looks like"
# # puts frames
# tframes = genome.translate_with_frame(dna)
# puts "The protiens read as 6 frames looks like" 
# puts tframes

File: genetic_algorith_python/bio_genome.py
import random

class Genome:
    def __init__(self):
        # universal amino acid dictionary
        self.aminoacid_dict = {
            'UUC':'F','UUU':'F','UUA':'L','UUG':'L','CUU':'L','CUC':'L','CUA':'L','CUG':'L',
            'AUU':'I','AUC':'I','AUA':'I','AUG':'M','GUU':'V','GUC':'V','GUA':'V','GUG':'V',
            'UCU':'S','UCC':'S','UCA':'S','UCG':'S','AGU':'S','AGC':'S','CCU':'P','CCC':'P','CCA':'P','CCG':'P',
            'ACU':'T','ACC':'T','ACA':'T','ACG':'T','GCU':'A','GCC':'A','GCA':'A','GCG':'A',
            'UAU':'Y','UAC':'Y','CAU':'H','CAC':'H','CAA':'Q','CAG':'Q',
            'AAU':'N','AAC':'N','AAA':'K','AAG':'K','GAU':'D','GAC':'D','GAA':'E','GAG':'E',
            'UGU':'C','UGC':'C','UGG':'W','CGU':'R','CGC':'R','CGA':'R','CGG':'R','AGA':'R','AGG':'R',
            'GGU':'G','GGC':'G','GGA':'G','GGG':'G',
            'UAA':'Stop','UGA':'Stop','UAG':'Stop'
        }

    def generate_ssDNA(self, length):
        """Generate a random single-stranded DNA sequence"""
        return ''.join(random.choice("ATGC") for _ in range(length))

    def generate_rna(self, length):
        """Generate a random RNA sequence"""
        return ''.join(random.choice("AUGC") for _ in range(length))

    def complementary_DNA_strand(self, dna):
        """Generate complementary strand"""
        mapping = str.maketrans("ATGC", "TACG")
        return dna.upper().translate(mapping)

    def check_DNA(self, seq1, seq2):
        """Check complementarity"""
        if len(seq1) >= len(seq2):
            strand2 = seq2[::-1]  # reverse
            strand2_check = seq1.translate(str.maketrans("ACTG", "TGAC"))
            return strand2 in strand2_check
        else:
            strand1 = seq1[::-1]
            strand1_check = seq2.translate(str.maketrans("ACTG", "TGAC"))
            return strand1 in strand1_check

    def six_reading_frames(self, seq1, rseq2):
        """Return six reading frames for DNA and its reverse complement"""
        output = []

        # forward frames
        f1 = ' '.join([seq1[i:i+3] for i in range(0, len(seq1), 3)])
        f2 = ' '.join([seq1[i:i+3] for i in range(1, len(seq1)-2, 3)])
        f3 = ' '.join([seq1[i:i+3] for i in range(2, len(seq1)-2, 3)])
        output.append("\n".join([f1, f2, f3]))

        # reverse complement frames
        r1 = ' '.join([rseq2[i:i+3] for i in range(0, len(rseq2), 3)])
        r2 = ' '.join([rseq2[i:i+3] for i in range(1, len(rseq2)-2, 3)])
        r3 = ' '.join([rseq2[i:i+3] for i in range(2, len(rseq2)-2, 3)])
        output.append("\n".join([r1, r2, r3]))

        return output

    def translate_with_frame(self, dna, frames=[1,2,3,-1,-2,-3]):
        """Translate DNA sequence into protein sequences for given frames"""
        if not dna:
            return ["" for _ in frames]

        def to_rna(seq): return seq.replace('T', 'U')

        forward = [to_rna(dna[i:]) for i in range(3)]
        reverse = [to_rna(self.complementary_DNA_strand(dna[::-1])[i:]) for i in range(3)]

        translations = []
        for seq in forward + reverse:
            codons = [seq[i:i+3] for i in range(0, len(seq)-2, 3)]
            aa_seq = ''.join(self.aminoacid_dict.get(c, '') for c in codons)
            translations.append(aa_seq)

        frame_map = {1:0, 2:1, 3:2, -1:3, -2:4, -3:5}
        return [translations[frame_map[f]] for f in frames]

    def dna_to_rna(self, dna):
        """Convert DNA to RNA"""
        return dna.replace("T", "U")

    def protein(self, rna):
        """Convert RNA to protein sequence"""
        codons = [rna[i:i+3] for i in range(0, len(rna)-2, 3)]
        return ''.join(self.aminoacid_dict.get(c, '') for c in codons)


# Example usage
if __name__ == "__main__":
    g = Genome()
    dna = g.generate_ssDNA(21)
    comp = g.complementary_DNA_strand(dna)
    print("DNA:", dna)
    print("Complement:", comp)
    print("Frames:", g.six_reading_frames(dna, comp))
    print("Translation:", g.translate_with_frame(dna))
    print("RNA:", g.dna_to_rna(dna))
    print("Protein:", g.protein(g.dna_to_rna(dna)))
